### DISCUSSION TITLE
How can I set a required field when a user is created with oauth2?

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1613] [Q&A]
I'm trying to set a custom required user field named `group` while oauth2 signin/signup is happening.  I can't seem to figure out which event to tap into to be able to set this field.  Obviously allowing the user to set their group would be bad.

Another possibility would be for me to create another collection for group membership that is created in the user post create hook, but it's just not how I want to do it.

I was able to confirm that the `group` property is what's catching this up by turning off `required` setting.

**Comment by dschissler** [comment]
@ganigeorgiev Is [this commit](https://github.com/pocketbase/pocketbase/commit/36ab3fd1627397fe323dae04a7e75122184e9bf1) going to help me out?

**Reply by ganigeorgiev** [reply]
Yes, the new event hooks in the `develop` branch could be used for that.

But depending on the desired workflow you can also just use the collection create rule to restrict/limit the user submitted fields (the create rule will be checked before user creation during the OAuth2 signup process), eg. something like:

```js
@request.data.group:isset = false

// or match with a fixed value
@request.data.group = "yourGroup"
```

**Reply by dschissler** [reply]
I just realized that I can set it with `createData` until the those new much more intricate events make into a release.

**Reply by dschissler** [reply]
I just had an idea though, it seems as though `OnModelBeforeCreate` for `users` isn't being reached because there is validation required happening before it reaches that point.  Shouldn't the `required` check happen somewhere between the Before and After hooks though?  It seems like the Before model hook is the place to set default values.

Is there any serious edge case or bad things that can happen if you allow user code to start playing with possibly invalid values to derive for other values?  In my case I just want to set `group` as `normal` value.  But if you gave access to me there then possibly I could take an unsafe unvalidated value for use in another field that could trigger a security issue for example.

This might seem annoying, but it's almost like to be able to use model hook only instead of a combination of model hooks and API request hooks that I'd need something like `OnModelBeforeCreate_BeforeValidation`, `OnModelBeforeCreate_AfterValidation`, and `OnModelAfterCreate`.

Said in another way, my problem is that I don't have access to `OnModelBeforeCreate` because the create action is being rejected since the validation is determining that a required field is empty.  The whole point of this hook for me is to fill in a required value (like what a normal framework would allow).

I don't like needing to place the group `normal` in Javascript, because it's then just another constant that I either need to hard code or to share from server to front end.

I also don't like how any record with a required field that is created needs to have an API event to set it, because it also needs to be set in Golang code in any additional place where a record is created in that collection.  I seems bad to me to have default values being defined outside of the model (hooks).  It's brittle and much harder to track down.

I also don't like that when these new events come out (that are in develop branch now) that for each auth record (like users) then the default value would need to be set in both the password and oauth2 authentication (since they're different events).  It's not clear how these two methods differ.

I understand that Pocketbase grew out of being a sort of smart but dumb repository with cool interface and so it was all supposed to happen externally, and then internal access (as a framework) became possible, and so sending a default value in wasn't really an issue since that is just how it needed to be done anyways.  It's kind of a problem for me though using it as a framework because so much is happening internally already.

Correct me if I'm wrong but basically the null and empty string sqlite woes led to no default values, and then the only solution was the client setting default values because that was all that Pocketbase was anyways.  Unfortunately now it's just painful for me.

Maybe I'm confused though.

**Reply by dschissler** [reply]
As I understand it, if currently I create a non-auth record with a required value then I can set that value in `OnRecordBeforeCreateRequest` and that if I didn't that it would already be rejected as invalid before it reached `OnModelBeforeCreate`.

Auth records are complicated by there being two ways to create them.  So after this new commit of yours will I need to set a default value in both `OnRecordBeforeCreateRequest` and `OnRecordBeforeAuthWithOAuth2Request`?  Is this going to work, or will it be rejected as invalid before it reaches either one of those?

I think that it could be done better somehow (probably in a model before hook that reaches before validation occurs), but I could live with calling out to a central hook location in both events.

**Reply by ganigeorgiev** [reply]
> Shouldn't the required check happen somewhere between the Before and After hooks though?

`OnModelBeforeCreate` and `OnModelBeforeUpdate` hooks are triggered at the end, right before the insert/update in the database.

> I also don't like that when these new events come out (that are in develop branch now) that for each auth record (like users) then the default value would need to be set in both the password and oauth2 authentication (since they're different events). It's not clear how these two methods differ.

The new `OnRecordBeforeAuthWithPasswordRequest` hook is triggered before password authentication. No user is created in this scenario (_unless you manually create it inside the hook in case the identity related record is missing; this is to support password+otherField authentications_).

The `OnRecordBeforeAuthWithOAuth2Request` could create a new user on OAuth2 sign-ups, aka. when there is no existing account to link the OAuth2 user (_the hook also allows you to specify a custom record to link in case you want to do that_).

> I understand that Pocketbase grew out of being a sort of smart but dumb repository with cool interface and so it was all supposed to happen externally, and then internal access (as a framework) became possible, and so sending a default value in wasn't really an issue since that is just how it needed to be done anyways.

PocketBase was designed to be used as framework and to be extendable since its initial design (see [Presentator #183](https://github.com/presentator/presentator/issues/183)). Yes, a lot of things could have been designed and structured better and some of the early design decisions will be improved but it will take time. Option to set default values was requested several times in the past (_https://github.com/pocketbase/pocketbase/issues/1043, https://github.com/pocketbase/pocketbase/issues/402, https://github.com/pocketbase/pocketbase/issues/206, and probably others_) but for me it is ambiguous and for now it is not planned and it is better left to be set client-side (or server-side in a hook).

> Correct me if I'm wrong but basically the null and empty string sqlite woes led to no default values, and then the only solution was the client setting default values because that was all that Pocketbase was anyways. Unfortunately now it's just painful for me.

I don't understand what do you mean. SQLite has nothing to do with PocketBase not having default values. We just don't store `null` as field value (_with the exeption of the `json` field type_) and therefore it is not clear when the defaults will be applied.

**Reply by dschissler** [reply]
Always, thanks for the thoughtful and ultra fast response.  Sometimes when I write a lot of technical complaints then people perceive it as yelling at them.  This is all complicated and as you said, it's complicated and not easily clear.

So let me highlight the solution that I came up with.  Bad naming aside, there would be a `OnModelBeforeCreate_BeforeValidation` and `OnModelBeforeCreate_AfterValidation` event.  So normally with a generic framework the user would have created a validation check sometime after setting default values in code.  Since you're handling that for us to a degree, then there needs to be a way to get in the middle of it.  I understand that you might be reluctant to create a huge amount of events, but by handling validation built-in then you're essentially denying us normal and typical usage of the before create and before update events.

**Reply by dschissler** [reply]
@ganigeorgiev What about creating a non-daos event registration on `app` so that `RecordUpsert` can call it before validation happens?  I'm assuming that API validation happens in `RecordUpsert`.  This is easy.

**Reply by ganigeorgiev** [reply]
Dao works only with the database. Validations happen inside the forms.
We can add form specific events but I don't want to rush it. I understand what is the issue and the inconvenience you are trying to solve, but I'll need to think a little more on it before adding new events.

For now the easiest solutions if you want to have a default `group` for both OAuth2 record and regular API creation would be:

1. Send the value client-side and validate the submitted value in the Create/Update API rules.
2. Remove the "Nonempty" constraint for the field and always set a default value in the `OnModelBeforeCreate/OnModelBeforeUpdate` hooks if it is empty or different than the expected value, eg.:
    ```go
    app.OnModelBeforeCreate().Add(func(e *core.ModelEvent) error {
        record, ok := e.Model.(*models.Record)
        if !ok || record.Collection().Name != "users" {
            return nil
        }

        if record.GetString("group") == "" {
            record.Set("group", "normal")
        }

        return nil
    })
    ```

**Reply by dschissler** [reply]
Thanks.  I can see that you perfectly understand the situation.  I'll take your advice for 2), except that I'll be removing the `if record.GetString("group") == "" {` check because that seems like a massive security hole since they could just set themselves to (normal user) admin there.

What if there was a `OnModelCreateBeforeValidation` event that was always called/registered on `dao`.  Then the upsert form validation was being sent into `dao` as a function in the create method.

So the dao events would be in the order;

1) `ModelCreateInit`, or `OnModelCreateBeforeValidation`
2) `ModelValidation` (assigned from the upsert form)
3) `ModelBeforeCreate`

So if `ModelValidation` wasn't assigned (when running dao directly) then it would just be skipped.  But the 1) function would always execute to fill in default values or to prepare somehow for validation.
